File: Python/SOLID/1. SingleResp.py
class Journal:
    def __init__(self):
        self.entries = []
        self.count = 0

    def addEntry(self, text):
        self.count += 1
        self.entries.append(f"{self.count}:{text}")

    def removeEntry(self, pos):
        del self.entries[pos]

    def __str__(self):
        return "\n".join(self.entries)

    # the below functionality might be used by other objects as well.
    # As persistence is sth employed by various obhjects we cannot have them all in every class
    """
    def save(self, filename):
        file = open(filename, "w")
        file.write(str(self))
        file.close()

    def load(self, filename):
        pass

    def low_from_web(self, uri):
        pass
    """


class PersistenceManager:
    def saveToFile(journal, filename):
        file = open(filename, "w")
        file.write(str(journal))
        file.close()


j = Journal()
j.addEntry("I cried today!")
j.addEntry("I ate a bug")
print(f"Journal entires: \n{j}")

file = r"journal.txt"
PersistenceManager.saveToFile(j, file)
with open(file) as fh:
    print(fh.read())

File: Javascript/SOLID/1. SingleResp.js
const fs = require("fs");

class Journal {
  constructor() {
    this.entries = {};
    this.count = 0;
  }

  addEntry(text) {
    let c = ++this.count;
    let entry = `${c}: ${text}`;
    this.entries[c] = entry;
    return c;
  }

  removeEntry(index) {
    delete this.entries[index - 1];
  }

  toString() {
    return Object.values(this.entries).join("\n");
  }

  // the below functionality might be used by other objects as well.
  // As persistence is sth employed by various obhjects we cannot have them all in every class
  /*
  save(filename) {
    fs.writeFileSync(filename, this.toString());
  }

  load(filename) {}
  loadFromWeb(url) {}
  */
}

class PersistenceManager {
  saveToFile(journal, filename) {
    fs.writeFileSync(filename, journal.toString());
  }
}

let j = new Journal();
j.addEntry("I cried today");
j.addEntry("I ate a bug");
console.log(j.toString());

let p = new PersistenceManager();
let filename = "journal.txt";
p.saveToFile(j, filename);
